### PR TITLE
EREGCSC-1783 -- default to Search.gov resources results on search page & style

### PR DESCRIPTION
**Pull request by PhilR8**
Resolves [EREGCSC-1783](https://jiraent.cms.gov/browse/EREGCSC-1783)

**Description**

When visiting the Search page, the Resources results (the right column) will return results from Search.gov.

These results will differ from the results on the Resources page.

**This pull request changes:**

- Adds [AJV](https://ajv.js.org/) schema validation library, used for Cypress tests that validate response schema for Search.gov resources endpoint
- Makes Search.gov resources endpoint (`/v3/resources/search`) the default endpoint for resources requests on the combined search page
- Creates `SearchGovResults` Vue component that is specific to the resources results on the combined search page
- Creates new `ResultItem` Vue component that is used for both Reg search results and Resources search results on the combined search page -- this is to keep the styles consistent for both sets of results
- Parses result `snippet` text to find terms that need to be highlighted, demarcated by special characters `&#xe000` and `&#xe001`.  Surrounds each term with a span tag with a style class so it is bolded.
   * Does the same thing for `description`, which also needs to have the search terms bolded if they are present.
- Removes `filters` from some Vue components in anticipation of the upgrade to Vue 3, which deprecates and removes template filters.  These filters have been converted to plain old JS functions and put into a `filters.js` file
- Removes styles from certain Vue single file components and puts them in the SCSS files.  It was probably a mistake to add styles to single file components; keeping all styles in one place in SCSS files is probably better for maintainability and to reduce accidental duplication etc.

**Steps to manually verify this change:**

1. Visit [experimental deployment](https://ut28at6gx0.execute-api.us-east-1.amazonaws.com/dev762) and search for certain things.  `Community First Choice` is a good string on which to begin testing.  Compare results to other experimental deployments or `dev`
2. Compare to Figma [here](https://www.figma.com/file/iZFh6UhvGVEHw1D4IAt3Bw/eRegs-Combined-Search?node-id=3908-11853&t=Jv9sSDLGdRMyagGp-0)

### CODE REVIEW

File: solution/ui/regulations/css/scss/partials/_search.scss [hunk 1]
```diff
@@ -5,164 +5,149 @@
...
+            }
         }
-    }
 
-    // Right Sidebar
+        .pagination-expand-row {
+            display: flex;
+            flex-direction: row;
+            justify-content: center;
+            margin-bottom: 100px;
```
Comment (PhilR8) [inline solution/ui/regulations/css/scss/partials/_search.scss:140]: some styles, like this one, aren't new -- they have been removed from the style tag of a Vue component and moved into a SCSS stylesheet.  These styles were moved from Search.vue to here.

File: solution/ui/regulations/eregs-vite/src/views/Search.vue [hunk 2]
```diff
@@ -474,105 +469,4 @@ export default {
 };
 </script>
 
-<style lang="scss">
-#searchApp.search-view {
-    display: flex;
-    flex-direction: column;
-
-    > .nav-container {
```
Comment (PhilR8) [inline solution/ui/regulations/eregs-vite/src/views/Search.vue:482]: Styles removed from Vue component and moved into SCSS stylesheet.

File: solution/ui/regulations/eregs-vite/src/utilities/filters.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    const raw_date = dateString.split("-");
+    if (raw_date.length > 1) {
+        options.month = "long";
+    }
+    if (raw_date.length > 2) {
+        options.day = "numeric";
+    }
+    const format = new Intl.DateTimeFormat("en-US", options);
+    return format.format(date);
+};
```
Comment (PhilR8) [inline solution/ui/regulations/eregs-vite/src/utilities/filters.js:13]: Template filters are no longer supported in Vue 3, so I'm moving some out into a global filters file and just importing them into components and using them as normal JS functions.

File: solution/ui/regulations/eregs-vite/src/components/search/ResultsItem.vue [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        </div>
+        <div class="result__link">
+            <slot name="link"></slot>
+        </div>
+        <div class="result__snippet">
+            <slot name="snippet"></slot>
+        </div>
+        <slot name="sections"></slot>
+    </div>
+</template>
```
Comment (PhilR8) [inline solution/ui/regulations/eregs-vite/src/components/search/ResultsItem.vue:17]: a little presentation component used for structure and styling only.

File: solution/ui/regulations/eregs-vite/src/components/resources/ResourcesResults.vue [hunk 5]
```diff
@@ -94,56 +96,7 @@ export default {
...
-    destroyed() {},
-
-    data() {
-        return {
-            dataProp: "value",
-        };
-    },
-
-    filters: {
-        locationLabel(value) {
```
Comment (PhilR8) [inline solution/ui/regulations/eregs-vite/src/components/resources/ResourcesResults.vue:120]: Template filters are no longer supported in Vue 3, so in anticipation of the eventual upgrade from Vue 2 to Vue 3, I'm moving some filters out into a global filters file and just importing them into components and using them as normal JS functions.

See more here: https://wildermuth.com/2020/05/11/Where-Did-Filters-Go-in-Vue-3/

File: solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            )
+        )
+    );
+};
+
+const formatLinkTitle = ({ description, descriptionHeadline, snippet }) => {
+    const termsToHighlight = getHighlightTerms(snippet, openChar, closeChar);
+
+    let linkTitle = descriptionHeadline || description;
+    termsToHighlight.forEach((term) => {
```
Comment (PhilR8) [inline solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue:63]: bolded search term matches are not included in the `description` in the response from the API, but the story AC mandates that the `description` must included the same bolded search terms as the `snippet`.  I'm doing that on the front end here -- collecting the highlighted search terms from the `snippet` and then finding them in the `description` and surrounding them with a `<span>` tag that includes a bolded style class.

File: solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        required: true,
+    },
+    results: {
+        type: Array,
+        default: () => [],
+    },
+});
+
+const openChar = ""; // &#xe000;
+const closeChar = ""; // &#xe001;
```
Comment (PhilR8) [inline solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue:31]: not sure how to target these characters via an explicit unicode/HTML code reference in JS.

File: solution/ui/regulations/eregs-vite/src/views/Search.vue [hunk 8]
```diff
@@ -351,12 +349,9 @@ export default {
         },
         async retrieveResourcesResults({ query, page, pageSize }) {
```
Comment (thwalker6) [inline solution/ui/regulations/eregs-vite/src/views/Search.vue:350]: I dont think we ever plan on changin this from the default of 50, can probably just remove it
```suggestion
        async retrieveResourcesResults({ query, page, pageSize }) {
```

File: solution/ui/regulations/eregs-vite/src/utilities/api.js [hunk 9]
```diff
@@ -416,12 +416,20 @@ const getSupplementalContent = async (
...
 
-const getCategories = async () =>  httpApiGet("resources/categories");
+/**
+ * @param {string} [page=1] - page number of paginated results to return.
+ * @param {string} [q=""] - search querystring
+ *
+ * @returns {Promise<{count: number, next: string, previous: string, results: Array<Object>}>} - Promise that contains response object when fulfilled
+ */
+const getSearchGovResources = async ({ page = 1, q = "" }) =>
+    httpApiGet(`resources/search?q=${q}&page=${page}`);
```
Comment (thwalker6) [inline solution/ui/regulations/eregs-vite/src/utilities/api.js:430]: This is a note, will need to update this line for uriencoding once my current PR is in.

File: solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    const reOpen = new RegExp(startChar, "g");
+    const reClose = new RegExp(stopChar, "g");
+
+    return snippet
+        .replace(reOpen, `<span class="${highlightClass}">`)
+        .replace(reClose, "</span>");
+};
+</script>
+
+<template>
```
Comment (thwalker6) [inline solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue:119]: I know this is technically fine but we typically have template first on our files and then functions and whatnot.  Is this a Vue3 style change or is it just how you did it on this file.  We should be consistent.

File: solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+const needsBar = (item) => item.date && item.name;
+
+/**
+ * @param string {string} - string containing substrings surrounded by special characters
+ * @param startChar {string} - character that indicates the start of a substring to get
+ * @param stopChar {string} - character that indicates the end of a substring to get
+ * @returns {Array{string}} - Array of unique substrings
+ */
+const getHighlightTerms = (string, startChar, stopChar) => {
```
Comment (thwalker6) [inline solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue:42]: nicely done here, took me a second to wrap my head around it

File: solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue (hunk 10 above)
Comment (PhilR8) [reply solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue:119]: Going forward using the Vue 2.7/3 Composition API, the official docs have `<script>` above `<template>`.  This is in line with how React does it with `jsx`/`tsx` files.

https://vuejs.org/guide/extras/composition-api-faq.html#what-is-composition-api

File: solution/ui/regulations/eregs-vite/src/views/Search.vue (hunk 8 above)
Comment (PhilR8) [reply solution/ui/regulations/eregs-vite/src/views/Search.vue:350]: there were talks about eventually adding the ability for users to select how many resources they'd like to see per page.  It's also functionality the SearchGov endpoint should have if we end up using it.  I'd rather keep it here for now.

File: solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue (hunk 11 above)
Comment (PhilR8) [reply solution/ui/regulations/eregs-vite/src/components/resources/SearchGovResults.vue:42]: I only got halfway to nirvana.  Next time I do something like this, I'm going to get a `pipe` operator working: https://jrsinclair.com/articles/2022/javascript-function-composition-whats-the-big-deal/

Or maybe the JS pipeline operator (`|>`) will be released with ES2023 by then.  
https://github.com/tc39/proposal-pipeline-operator

File: solution/ui/regulations/eregs-vite/src/views/Search.vue (hunk 8 above)
Comment (PhilR8) [reply solution/ui/regulations/eregs-vite/src/views/Search.vue:351]: we've set the existing Django resources API endpoint back to the default call when requesting resources.  We've also re-enabled the ability to add `searchgov` as a URL parameter to request resources from the new SearchGov endpoint.  So `pageSize` is used again.

File: solution/ui/regulations/eregs-vite/src/utilities/api.js (hunk 9 above)
Comment (PhilR8) [reply solution/ui/regulations/eregs-vite/src/utilities/api.js]: This has been updated

**Review by thwalker6 — APPROVED** [review]
LGTM

**Pull request merged**